File: src/main/java/ru/arc/hooks/ztranslator/TranslatorHook.java
package ru.arc.hooks.ztranslator;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import ru.arc.ARC;

import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.warn;

public class TranslatorHook {

    Map<String, String> map = new HashMap<>();
    Gson gson = new Gson();

    public TranslatorHook() {
        Path path = ARC.plugin.getDataPath().resolve("lang.json");
        if (!Files.exists(path)) {
            warn("lang.json is not present in ARC folder! Disabling translating materials...");
            return;
        }
        try {
            String s = Files.readString(path);
            TypeToken<Map<String, String>> token = new TypeToken<>() {
            };
            map = gson.fromJson(s, token);
        } catch (Exception e) {
            error("Error loading translations", e);
        }
    }

    public String translate(ItemStack stack) {
        if (stack == null) return "Null";
        return map.getOrDefault(stack.translationKey(), toReadableName(stack.getType().name()));
    }

    public String translate(Material material) {
        return map.getOrDefault(material.translationKey(), toReadableName(material.name()));
    }

    private String toReadableName(String name) {
        StringBuilder sb = new StringBuilder();
        boolean start = true;
        for (char c : name.toCharArray()) {
            if (start) {
                sb.append(Character.toUpperCase(c));
                start = false;
            }else if (c == '_') {
                sb.append(" ");
                start = true;
            } else {
                sb.append(Character.toLowerCase(c));
            }
        }
        return sb.toString();
    }
}

File: src/main/kotlin/ru/arc/hooks/ztranslator/TranslatorHook.kt
package ru.arc.hooks.ztranslator

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import ru.arc.ARC
import ru.arc.util.Logging.error
import ru.arc.util.Logging.warn
import java.nio.file.Files

class TranslatorHook {

    private val map: Map<String, String>
    private val gson = Gson()

    init {
        val path = ARC.instance.dataPath.resolve("lang.json")
        if (!Files.exists(path)) {
            warn("lang.json is not present in ARC folder! Disabling translating materials...")
            map = emptyMap()
        } else {
            var loaded: Map<String, String> = emptyMap()
            try {
                val s = Files.readString(path)
                val token = object : TypeToken<Map<String, String>>() {}
                loaded = gson.fromJson(s, token)
            } catch (e: Exception) {
                error("Error loading translations", e)
            }
            map = loaded
        }
    }

    fun translate(stack: ItemStack?): String {
        if (stack == null) return "Null"
        return map.getOrDefault(stack.translationKey(), toReadableName(stack.type.name))
    }

    fun translate(material: Material): String =
        map.getOrDefault(material.translationKey(), toReadableName(material.name))

    private fun toReadableName(name: String): String {
        val sb = StringBuilder()
        var start = true
        for (c in name) {
            when {
                start -> { sb.append(c.uppercaseChar()); start = false }
                c == '_' -> { sb.append(' '); start = true }
                else -> sb.append(c.lowercaseChar())
            }
        }
        return sb.toString()
    }
}
